### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): Being an isomorphism is local at the target.

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+lemma isomorphisms_eq_isOpenImmersion_inf_surjective :
+    isomorphisms Scheme = (@IsOpenImmersion ⊓ @Surjective : MorphismProperty Scheme) := by
+  ext
+  exact (isIso_iff_isOpenImmersion _).trans
+    (and_congr Iff.rfl ((TopCat.epi_iff_surjective _).trans (surjective_iff _).symm))
+
+lemma isomorphisms_eq_stalkwise :
+    isomorphisms Scheme = (isomorphisms TopCat).inverseImage Scheme.forgetToTop ⊓
+      stalkwise (fun f ↦ Function.Bijective f) := by
+  rw [isomorphisms_eq_isOpenImmersion_inf_surjective, isOpenImmersion_eq_inf,
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean:28]: I think we should have `IsHomeomorphism` and then state this in terms of it, i.e. `isomorphisms Scheme = topologically IsHomeomorphism ⊓ stalkwise (fun f ↦ Function.Bijective f)`.

File: Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+section Surjective
+
+variable {X Y Z : Scheme.{u}} (f : X ⟶ Y) (g : Y ⟶ Z)
+
+/-- A morphism of schemes is surjective if the underlying map is. -/
+@[mk_iff]
+class Surjective : Prop where
+  surj : Function.Surjective f.1.base
+
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean:56]: I am wondering if we should agree on a convention, whether we write `f.val.base` or `f.1.base` or even `f.1.1`. I personally prefer the first version as I think it is the most readable one.

File: Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+of the underlying map, including
+
+- `Injective`
+- `Surjective`
+- `IsOpenMap`
+- `IsClosedMap`
+- `Embedding`
+- `OpenEmbedding`
+- `ClosedEmbedding`
+
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean:23]:
```suggestion
```

File: Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+-/
+import Mathlib.Topology.LocalAtTarget
+import Mathlib.AlgebraicGeometry.Morphisms.Constructors
+
+/-!
+
+## Properties on the underlying functions of morphisms of schemes.
+
+This file includes various results on properties of morphisms of schemes that come from properties
+of the underlying map, including
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean:14]:
```suggestion
of the underlying map of topological spaces, including
```

File: Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+lemma Surjective.comp_iff [Surjective f] : Surjective (f ≫ g) ↔ Surjective g :=
+  ⟨fun _ ↦ of_comp f g, fun _ ↦ inferInstance⟩
+
+instance : MorphismProperty.RespectsIso @Surjective :=
+  surjective_eq_topologically ▸ topologically_respectsIso _ (fun e ↦ e.surjective)
+    (fun _ _ hf hg ↦ hg.comp hf)
+
+instance surjective_isLocalAtTarget : IsLocalAtTarget @Surjective := by
+  have := AlgebraicGeometry.instRespectsIsoSchemeSurjective
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean:78]:
```suggestion
  have : MorphismProperty.RespectsIso @Surjective := inferInstance
```
I don't know how stable it is to rely on the automatically generated instance name.

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean:78]: (alternatively: give the `RespectsIso` instance a name)

File: Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean:28]: Hmmm I think `isomorphisms` is exactly this and already has some useful API around it.

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean:28]: But in terms of category theory which does not always play nicely with the type theoretic topology library.

File: Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean (hunk 2 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean:55]: I personally think spelling it out is a good thing but mathlib (or at least I) uses `.1` and `.val` interchangeably so I end up using `f.1.base` most of the time. I don't think the vague `val` gives much information that is worth +2 chars but I'm not against a library-wide replacement if you feel like it.

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean:55]: At least `.val` indicates that it is a subtype where `.1` can really be anything. In any case, this is not important for this PR.

File: Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean:28]: There is #12533 but it seems like a stale PR.

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean:28]: Hm that's sad. Okay, let's merge this one as is and refactor later (or just add the suggested formulation) when we have `IsHomeomorphism`.

**Review by chrisflav — APPROVED** [review]
Thanks! LGTM

File: Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/IsIso.lean:28]: I've asked on the relevant Zulip thread whether the author intend to continue and would probably take over if the author has decided to abandon it.
I'll add the suggestion once that PR is merged. I think this lemma is still useful though and I'll probably still keep this.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**